testsuite/integration/manualmode/src/test/java/org/jboss/as/test/manualmode/undertow/CustomUndertowFilterTestCase.java

```diff
@@ -0,0 +1,206 @@
+ *
+ */
+package org.jboss.as.test.manualmode.undertow;
+
+import org.apache.http.Header;
+import org.apache.http.HttpResponse;
+import org.apache.http.StatusLine;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.jboss.as.test.integration.management.util.CLIWrapper;
+import org.jboss.as.test.integration.management.util.ServerReload;
+import org.jboss.as.test.shared.TestSuiteEnvironment;
+import org.jboss.logging.Logger;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.exporter.ZipExporter;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.wildfly.core.testrunner.ServerControl;
+import org.wildfly.core.testrunner.ServerController;
+import org.wildfly.core.testrunner.WildflyTestRunner;
+
+import javax.inject.Inject;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Tests custom http handler(s) configured as {@code custom-filter}s in the undertow subsystem
+ *
+ * @author Jaikiran Pai
+ */
+@RunWith(WildflyTestRunner.class)
+@ServerControl(manual = true)
+public class CustomUndertowFilterTestCase {
+
+    private static final Logger logger = Logger.getLogger(CustomUndertowFilterTestCase.class);
+
+    private static final String CUSTOM_FILTER_MODULE_NAME = "custom-undertow-filter-module";
+    private static final String CUSTOM_FILTER_CLASSNAME = CustomHttpHandler.class.getName();
+    private static final String CUSTOM_FILTER_RESOURCE_NAME = "testcase-added-custom-undertow-filter";
+    private static final String WAR_DEPLOYMENT_NAME = "test-tccl-in-custom-undertow-handler-construction";
+
+    @Inject
+    private static ServerController CONTROLLER;
+
+    @BeforeClass
+    public static void setupServer() throws Exception {
+        CONTROLLER.start();
+        // setup the server with the necessary Undertow filter configurations
+        prepareServerConfiguration();
+        // deploy an web application
+        deploy();
+    }
+
+    @AfterClass
+    public static void resetServer() throws Exception {
+        try {
+            undeploy();
+        } catch (Exception e) {
+            // ignore
+            logger.debug("Ignoring exception that occurred during un-deploying", e);
+        }
+        try {
+            // cleanup the undertow configurations we did in this test
+            resetServerConfiguration();
+        } finally {
+            CONTROLLER.stop();
+        }
+    }
+
+    private static void prepareServerConfiguration() throws Exception {
+        try (final CLIWrapper cli = new CLIWrapper(true)) {
```

Comment:
I am not really big fan of setting up and using CLI here. CLI starts up new java process and uses remoting to communicate back.

for creating module itself we do have org.jboss.as.test.module.util.TestModule
and for mgmt. opts, it is better to just use ManagementClient (via CONTROLLER#getClient)

Reply:
>> CLI starts up new java process and uses remoting to communicate back.

I wasn't aware of that! I borrowed most of this management util code from some other tests. I'll fix accordingly. Thank you for letting me know.